File: introHPC/10/shell.py
import zmq

context = zmq.Context()

sender = context.socket(zmq.PUSH)
sender.bind("tcp://*:6666")

receiver = context.socket(zmq.PULL)
receiver.bind("tcp://*:7777")

while (True):
    sum = 0
    max = int(input(">>> "))
    for i in range(1, max+1):
        sender.send_string(b'%s' % i)
        print("Sent: %s" % i)
    print("data sent.")
    for i in range(1, max+1):
        num = int(receiver.recv_string())
        print("Received: %s" % num)
        sum = sum + num
    print("\nAnswer is : %s" % sum)

File: introHPC/10/shell.cpp
#include <zmqpp/zmqpp.hpp>
#include <iostream>

auto main() -> int
{
    zmqpp::context context;

    zmqpp::socket sender(context, zmqpp::socket_type::push);
    sender.bind("tcp://*:6666");

    zmqpp::socket receiver(context, zmqpp::socket_type::pull);
    receiver.bind("tcp://*:7777");

    while (1) {
        int sum = 0;
        int max;
        std::cout << ">>>";
        std::cin >> max;
        for (int i=1; i<=max; i++) {
            zmqpp::message message;
            message << std::to_string(i);
            sender.send(message);
            std::cout << "Sent: " << i << std::endl;
        }
        std::cout << "Data sent." << std::endl;
        for (int i=1; i<=max; i++) {
            zmqpp::message message;
            std::string str;
            receiver.receive(message);
            message >> str;
            int num = std::stoi(str);
            sum += num;
            std::cout << "Received: " << num << std::endl;
        }
        std::cout << std::endl << "Answer is " << sum << std::endl;

    }
    return 0;
}
